<a href="https://colab.research.google.com/github/nestormontano/PrDS_2023_01_WIDS_Prediccion_Energia/blob/main/Scripts/WIDS2022_Prediccion_Energia__EDA_py.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Proyecto Programa Data Science WIDS - Prediccion de Energia

Con Python

Sitio del concurso: https://www.kaggle.com/c/widsdatathon2022

Datos: https://drive.google.com/drive/folders/1BxbUc5k1DowoyIvRurX1Iw8ksyIrLzBE?usp=share_link


Deben seguir las instrucciones dadas en el pdf, esto para  
- Ser invitados como colaboradores del repositorio en GitHub
- Abrir la carpeta y hacer un acceso directo en su drive 
- Conectar el Colab con su drive 

Recuerden que deben cambiar la carpeta que será el workspace o las rutas para poder importar los datos.

Finalmente, seguir las instrucciones del pdf para escribir en GitHub los cambios para que se noten sus aportes.

In [4]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


#### Preliminares

In [5]:
import pandas as pd
import numpy as np
import os
import textwrap

In [6]:
! pip install skimpy

  Attempting uninstall: rich
    Found existing installation: rich 13.3.3
    Uninstalling rich-13.3.3:
      Successfully uninstalled rich-13.3.3
  Attempting uninstall: ipykernel
    Found existing installation: ipykernel 5.5.6
    Uninstalling ipykernel-5.5.6:
      Successfully uninstalled ipykernel-5.5.6
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
google-colab 1.0.0 requires ipykernel~=5.5.6, but you have ipykernel 6.22.0 which is incompatible.


In [ ]:
!pip install -U dataprep

In [ ]:
from skimpy import skim
from dataprep.eda import create_report

In [11]:
os.chdir('/content/drive/MyDrive/8ed SEE')

In [12]:
''' Esta funcion es una utilitaria para imprimir los nombres de columnas y 
que no toque hacer desplazamiento horizontal'''
wrapper = textwrap.TextWrapper(width=100)

### Importar, corregir Datos y unir

## EDA

In [13]:
train= pd.read_csv('widsdatathon2022/train.csv')
train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 75757 entries, 0 to 75756
Data columns (total 64 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Year_Factor                75757 non-null  int64  
 1   State_Factor               75757 non-null  object 
 2   building_class             75757 non-null  object 
 3   facility_type              75757 non-null  object 
 4   floor_area                 75757 non-null  float64
 5   year_built                 73920 non-null  float64
 6   energy_star_rating         49048 non-null  float64
 7   ELEVATION                  75757 non-null  float64
 8   january_min_temp           75757 non-null  int64  
 9   january_avg_temp           75757 non-null  float64
 10  january_max_temp           75757 non-null  int64  
 11  february_min_temp          75757 non-null  int64  
 12  february_avg_temp          75757 non-null  float64
 13  february_max_temp          75757 non-null  int

In [14]:
# list( train.select_dtypes( include= 'integer') )  # Comando base
print( 'Columnas Object' )
print( wrapper.fill( str( list( train.select_dtypes( include= 'object') ) ) ) )
print( 'Columnas Entero' )
print( wrapper.fill( str( list( train.select_dtypes( include= 'integer') ) ) ) )
print( 'Columnas Float' )
print( wrapper.fill( str( list( train.select_dtypes( include= 'floating') ) ) ) )
print( 'Columnas Verdadero Falso' )
print( wrapper.fill( str( list( train.select_dtypes( include= 'bool') ) ) ) )
print( 'Columnas Categoricas' )
print( wrapper.fill( str( list( train.select_dtypes( include= 'category') ) ) ) )
print( 'Columnas Fecha' )
print( wrapper.fill( str( list( train.select_dtypes( include= 'datetime') ) ) ) )

Columnas Object
['State_Factor', 'building_class', 'facility_type']
Columnas Entero
['Year_Factor', 'january_min_temp', 'january_max_temp', 'february_min_temp', 'february_max_temp',
'march_min_temp', 'march_max_temp', 'april_min_temp', 'april_max_temp', 'may_min_temp',
'may_max_temp', 'june_min_temp', 'june_max_temp', 'july_min_temp', 'july_max_temp',
'august_min_temp', 'august_max_temp', 'september_min_temp', 'september_max_temp',
'october_min_temp', 'october_max_temp', 'november_min_temp', 'november_max_temp',
'december_min_temp', 'december_max_temp', 'cooling_degree_days', 'heating_degree_days',
'snowdepth_inches', 'days_below_30F', 'days_below_20F', 'days_below_10F', 'days_below_0F',
'days_above_80F', 'days_above_90F', 'days_above_100F', 'days_above_110F', 'id']
Columnas Float
['floor_area', 'year_built', 'energy_star_rating', 'ELEVATION', 'january_avg_temp',
'february_avg_temp', 'march_avg_temp', 'april_avg_temp', 'may_avg_temp', 'june_avg_temp',
'july_avg_temp', 'august_avg_temp'

###########################
**MIGUEL ZAMORA**
# Observemos que:
- id: variable str (identificación del  edificio).
- Year_Factor: variable categórica (año anonimizado)
Esta variable representa el año anonimizado en el que se observaron las condiciones climáticas y el consumo de energía del edificio. Dado que no se conocen los años reales y esta variable se utiliza para agrupar las observaciones, debemos tratarla como una variable categórica, pero la pondré como str.

Entonces, debemos convertir **id** y **Year_Factor** en string.

In [15]:
###############################################
# Miguel Zamora
train['id'] = train['id'].astype(str)
train['Year_Factor'] = train['Year_Factor'].astype(str)
###############################################

Analizar si los tipos de datos son los correctos, caso contrario modificarlos (ver link que les pasé al grupo para ejemplo)

In [ ]:
train['id'].dtype


dtype('int64')

In [ ]:
#Grupo AJK2 considero que el ID debería ser un string.
train['id'] = train['id'].astype('str')
train['id'].dtype

dtype('O')

In [16]:
# list( train.select_dtypes( include= 'integer') )  # Comando base
print( 'Columnas Object' )
print( wrapper.fill( str( list( train.select_dtypes( include= 'object') ) ) ) )
print( 'Columnas Entero' )
print( wrapper.fill( str( list( train.select_dtypes( include= 'integer') ) ) ) )
print( 'Columnas Float' )
print( wrapper.fill( str( list( train.select_dtypes( include= 'floating') ) ) ) )
print( 'Columnas Verdadero Falso' )
print( wrapper.fill( str( list( train.select_dtypes( include= 'bool') ) ) ) )
print( 'Columnas Categoricas' )
print( wrapper.fill( str( list( train.select_dtypes( include= 'category') ) ) ) )
print( 'Columnas Fecha' )
print( wrapper.fill( str( list( train.select_dtypes( include= 'datetime') ) ) ) )

Columnas Object
['Year_Factor', 'State_Factor', 'building_class', 'facility_type', 'id']
Columnas Entero
['january_min_temp', 'january_max_temp', 'february_min_temp', 'february_max_temp', 'march_min_temp',
'march_max_temp', 'april_min_temp', 'april_max_temp', 'may_min_temp', 'may_max_temp',
'june_min_temp', 'june_max_temp', 'july_min_temp', 'july_max_temp', 'august_min_temp',
'august_max_temp', 'september_min_temp', 'september_max_temp', 'october_min_temp',
'october_max_temp', 'november_min_temp', 'november_max_temp', 'december_min_temp',
'december_max_temp', 'cooling_degree_days', 'heating_degree_days', 'snowdepth_inches',
'days_below_30F', 'days_below_20F', 'days_below_10F', 'days_below_0F', 'days_above_80F',
'days_above_90F', 'days_above_100F', 'days_above_110F']
Columnas Float
['floor_area', 'year_built', 'energy_star_rating', 'ELEVATION', 'january_avg_temp',
'february_avg_temp', 'march_avg_temp', 'april_avg_temp', 'may_avg_temp', 'june_avg_temp',
'july_avg_temp', 'august_avg_temp'

### EDA univariado

Para encontrar qué variables deben ser excluidas del modelo, si se requiere alguna transformación, etc.

In [ ]:
skim(train)

In [ ]:
create_report(train)

#### ++ Variables con errores

Del eda univariado en lote, ¿existe alguna variable a la que le vean algo raro? Corregirlo (ojo, no digo que exista, deben analizar las tablas y gráficos a detalle)

In [ ]:
np.nan

nan

In [ ]:
train.loc[ train.max_wind_speed==0]

Empty DataFrame
Columns: [Year_Factor, State_Factor, building_class, facility_type, floor_area, year_built, energy_star_rating, ELEVATION, january_min_temp, january_avg_temp, january_max_temp, february_min_temp, february_avg_temp, february_max_temp, march_min_temp, march_avg_temp, march_max_temp, april_min_temp, april_avg_temp, april_max_temp, may_min_temp, may_avg_temp, may_max_temp, june_min_temp, june_avg_temp, june_max_temp, july_min_temp, july_avg_temp, july_max_temp, august_min_temp, august_avg_temp, august_max_temp, september_min_temp, september_avg_temp, september_max_temp, october_min_temp, october_avg_temp, october_max_temp, november_min_temp, november_avg_temp, november_max_temp, december_min_temp, december_avg_temp, december_max_temp, cooling_degree_days, heating_degree_days, precipitation_inches, snowfall_inches, snowdepth_inches, avg_temp, days_below_30F, days_below_20F, days_below_10F, days_below_0F, days_above_80F, days_above_90F, days_above_100F, days_above_110F, direction_max_wind_speed, direction_peak_wind_speed, max_wind_speed, days_with_fog, site_eui, id]
Index: []

[0 rows x 64 columns]

In [ ]:
# La variable year_built tiene ceros, hay que cambiar los ceros por NAN --- Por: GRUPO NM
train.loc[ train.year_built==0, 'year_built'] = np.nan

In [ ]:
train[['year_built']].describe()

year_built
count  73914.000000
mean    1952.465243
std       32.613687
min     1600.000000
25%     1927.000000
50%     1951.000000
75%     1977.000000
max     2015.000000

#### ++ Variables con demasiados NA

In [ ]:
porc_nas= train.isnull().mean()
porc_nas

Year_Factor                  0.000000
State_Factor                 0.000000
building_class               0.000000
facility_type                0.000000
floor_area                   0.000000
                               ...   
direction_peak_wind_speed    0.551909
max_wind_speed               0.542287
days_with_fog                0.604512
site_eui                     0.000000
id                           0.000000
Length: 64, dtype: float64

In [ ]:
porc_nas.max()

0.6045117942896366

In [ ]:
## Seleccionar y mostrar columnas con muchos NAs
#mínimo energy_star_rating 35% Grupo JAK2
print( wrapper.fill( str( 
    list( train.columns[ porc_nas > 0.35 ] )
      ) ) )

['energy_star_rating', 'direction_max_wind_speed', 'direction_peak_wind_speed', 'max_wind_speed',
'days_with_fog']


In [ ]:
############################################################################################
# Miguel Zamora
# Crear df con los valores faltantes indicando el porcentaje respecto del total de sus valores

# Obtener el total de valores faltantes en cada variable
total_faltante = train.isnull().sum()


# Calcular el porcentaje de cada valor faltante respecto al total de cada variable
porcentaje_faltante = total_faltante / len(train) * 100

# Crear la tabla de frecuencia de valores faltantes
tabla_faltantes = pd.DataFrame({'Total faltante': total_faltante, '% Faltante': porcentaje_faltante})

# Filtrar la tabla para mostrar solo las variables con valores faltantes
tabla_faltantes = tabla_faltantes[tabla_faltantes['Total faltante'] > 0]

# Mostrar la tabla de frecuencia
print(tabla_faltantes)


Los valores faltantes en estas variables pueden afectar la capacidad del modelo de predicción para predecir con precisión el consumo de energía del edificio. Es importante considerar el impacto de los valores faltantes en el análisis y la predicción del consumo de energía del edificio y tomar decisiones informadas sobre cómo manejarlos.

# Con esta información podemos concluir que:

- Las variables relacionadas con la velocidad y dirección del viento: **direction_max_wind_speed**, **direction_peak_wind_speed** y **max_wind_speed**) y la variable **days_with_fog** tienen un porcentaje significativo de datos faltantes, lo que podría afectar la calidad de las predicciones al usar estas variables en un modelo de aprendizaje automático.

- La variable **energy_star_rating** también tiene una cantidad considerable de datos faltantes (35.26%), lo que también podría afectar la calidad de las predicciones.

- La variable **year_built** tiene un porcentaje menor de datos faltantes (2.42%), pero aún podría ser necesario abordar estos valores faltantes antes de utilizar esta variable en un modelo de aprendizaje automático.

# Considerando la definición de las variables:

- **1. year_built:** 
Año en que se construyó el edificio.

- **2. energy_star_rating:** La calificación de eficiencia energética del edificio.

- **3. direction_max_wind_speed:** Dirección del viento para la velocidad máxima del viento en la ubicación del edificio, dada en direcciones de puntos cardinales de 360 grados.

- **4. direction_peak_wind_speed:** Dirección del viento para la velocidad máxima de ráfaga de viento en la ubicación del edificio, dada en direcciones de puntos cardinales de 360 grados.

- **max_wind_speed:** velocidad máxima del viento en la ubicación del edificio.

- **days_with_fog:** número de días con niebla en la ubicación del edificio.

# Esta información nos dice:

- El año de construcción de los edificios, que podría ser un factor que influye en la eficiencia energética, no estaba disponible para el 2.42% de los registros. Esto podría plantear desafíos al momento de evaluar cómo la antigüedad de un edificio impacta en su consumo de energía y cómo ha evolucionado la eficiencia energética a lo largo del tiempo.

- Además, la calificación Energy Star, que es un indicador clave de la eficiencia energética de un edificio, no estaba disponible en el 35.26% de los casos. Este dato faltante dificulta la comprensión del desempeño energético de un edificio y cómo las medidas de eficiencia podrían estar relacionadas con el consumo de energía.

- En cuanto a las condiciones climáticas, las direcciones de la velocidad máxima del viento y de la ráfaga máxima de viento, así como la velocidad máxima del viento, no estaban disponibles en más del 54% de los registros. Estos datos faltantes podrían afectar la capacidad de evaluar cómo el clima local, especialmente el viento, influye en el consumo de energía de un edificio.

- Por último, el número de días con niebla, que podría estar relacionado con la humedad y afectar los sistemas de climatización, no estaba disponible en el 60.45% de los registros. Esta falta de información complica la tarea de determinar si la presencia de niebla tiene un impacto significativo en el consumo de energía de los edificios.

# Basándonos en esa información, surgen algunas preguntas:

- **1.** ¿Cómo afecta la antigüedad de un edificio (año de construcción) a su eficiencia energética y al consumo de energía?

- **2.** ¿Existe una correlación entre la calificación Energy Star de un edificio y su Site EUI?

- **3.** ¿Cómo influyen las condiciones climáticas, como la velocidad y dirección del viento, y la cantidad de días con niebla, en el consumo de energía de los edificios?

- **4.** ¿Existen patrones geográficos en la eficiencia energética y el consumo de energía de los edificios en diferentes estados?

- **5.** ¿Cómo afectan otras características del edificio, como el tipo de instalación y el área del piso, al consumo de energía?


**Miguel Zamora**
##############################################################

**Debate**  

¿Cuál de estas variables debería eliminarse? Todas? O quizá exista alguna variable donde Y tenga un comportamiento diferente cuando esta llena vs cuando está con NA?? Existen variables cuyos NA son excluyentes (es decir, cuando la una tiene NA la otra tiene valor y viceversa)?

In [ ]:
pip install pingouin

In [ ]:
#No podemos eliminar la variable energy_star_rating porque el estudio se trata de energía
#Podemos eliminar la variables days_with_fog ya que cuando hay fog la energía es baja (Hay que averiguarlo)
#Grupo JAK2
import statsmodels.api as sm
import pingouin as pg
from scipy import stats
from scipy.stats import pearsonr

In [ ]:
#Grupo JAK2
display(pg.corr(train['energy_star_rating'], train['direction_max_wind_speed'], method='pearson'))
display(pg.corr(train['energy_star_rating'], train['direction_max_wind_speed'], method='spearman'))
display(pg.corr(train['energy_star_rating'], train['direction_max_wind_speed'], method='kendall'))

n         r          CI95%     p-val   BF10     power
pearson  21804 -0.000992  [-0.01, 0.01]  0.883594  0.009  0.052459

n         r          CI95%     p-val     power
spearman  21804  0.004587  [-0.01, 0.02]  0.498263  0.103972

n         r          CI95%     p-val     power
kendall  21804  0.003591  [-0.01, 0.02]  0.502361  0.082784

In [ ]:
#Grupo JAK2
display(pg.corr(train['energy_star_rating'], train['direction_peak_wind_speed'], method='pearson'))
display(pg.corr(train['energy_star_rating'], train['direction_peak_wind_speed'], method='spearman'))
display(pg.corr(train['energy_star_rating'], train['direction_peak_wind_speed'], method='kendall'))

n         r          CI95%     p-val   BF10     power
pearson  21223  0.002014  [-0.01, 0.02]  0.769255  0.009  0.059915

n         r          CI95%     p-val     power
spearman  21223  0.006163  [-0.01, 0.02]  0.369259  0.146229

n         r          CI95%     p-val     power
kendall  21223  0.004863  [-0.01, 0.02]  0.372806  0.109183

In [ ]:
#Grupo JAK2
display(pg.corr(train['energy_star_rating'], train['max_wind_speed'], method='pearson'))
display(pg.corr(train['energy_star_rating'], train['max_wind_speed'], method='spearman'))
display(pg.corr(train['energy_star_rating'], train['max_wind_speed'], method='kendall'))

n         r          CI95%     p-val   BF10     power
pearson  21804 -0.000333  [-0.01, 0.01]  0.960797  0.008  0.050276

n         r          CI95%     p-val     power
spearman  21804 -0.001848  [-0.02, 0.01]  0.785012  0.058567

n        r          CI95%     p-val     power
kendall  21804 -0.00147  [-0.01, 0.01]  0.783183  0.055411

In [ ]:
#Grupo JAK2
display(pg.corr(train['energy_star_rating'], train['days_with_fog'], method='pearson'))
display(pg.corr(train['energy_star_rating'], train['days_with_fog'], method='spearman'))
display(pg.corr(train['energy_star_rating'], train['days_with_fog'], method='kendall'))

n         r        CI95%     p-val   BF10     power
pearson  17823  0.019387  [0.0, 0.03]  0.009646  0.267  0.735141

n        r          CI95%     p-val     power
spearman  17823  0.00719  [-0.01, 0.02]  0.337115  0.160389

n         r          CI95%     p-val     power
kendall  17823  0.004759  [-0.01, 0.02]  0.366022  0.097362

In [ ]:
#Grupo JAK2
display(pg.corr(train['energy_star_rating'], train['avg_temp'], method='pearson'))
display(pg.corr(train['energy_star_rating'], train['avg_temp'], method='spearman'))
display(pg.corr(train['energy_star_rating'], train['avg_tembp'], method='kendall'))

n         r        CI95%         p-val       BF10  power
pearson  49048  0.086425  [0.08, 0.1]  5.892009e-82  3.963e+77    1.0

n         r         CI95%         p-val  power
spearman  49048  0.086065  [0.08, 0.09]  2.744365e-81    1.0

n         r         CI95%         p-val  power
kendall  49048  0.060685  [0.05, 0.07]  4.950271e-81    1.0

In [ ]:
#Grupo JAK2
#tomamos de ejemplo otra variable (avg_temp) para revisar la correlación
#pvalue = 0    cuando está por encima de cero podemos considerar la variable
#Eliminamos las variables
train.drop(columns=['direction_max_wind_speed', 'direction_peak_wind_speed', 'max_wind_speed', 'days_with_fog'])

Year_Factor State_Factor building_class  \
0                1      State_1     Commercial   
1                1      State_1     Commercial   
2                1      State_1     Commercial   
3                1      State_1     Commercial   
4                1      State_1     Commercial   
...            ...          ...            ...   
75752            6     State_11     Commercial   
75753            6     State_11    Residential   
75754            6     State_11     Commercial   
75755            6     State_11     Commercial   
75756            6     State_11    Residential   

                                   facility_type  floor_area  year_built  \
0                   Grocery_store_or_food_market     61242.0      1942.0   
1      Warehouse_Distribution_or_Shipping_center    274000.0      1955.0   
2                           Retail_Enclosed_mall    280025.0      1951.0   
3                      Education_Other_classroom     55325.0      1980.0   
4                      Warehouse_Nonrefrigerated     66000.0      1985.0   
...                                          ...         ...         ...   
75752                       Office_Uncategorized     20410.0      1995.0   
75753                        5plus_Unit_Building     40489.0      1910.0   
75754                           Commercial_Other     28072.0      1917.0   
75755                           Commercial_Other     53575.0      2012.0   
75756                         2to4_Unit_Building     23888.0      1974.0   

       energy_star_rating  ELEVATION  january_min_temp  january_avg_temp  ...  \
0                    11.0        2.4                36         50.500000  ...   
1                    45.0        1.8                36         50.500000  ...   
2                    97.0        1.8                36         50.500000  ...   
3                    46.0        1.8                36         50.500000  ...   
4                   100.0        2.4                36         50.500000  ...   
...                   ...        ...               ...               ...  ...   
75752                 8.0       36.6                28         43.451613  ...   
75753                98.0       36.6                28         43.451613  ...   
75754                 NaN       36.6                26         36.612903  ...   
75755                 NaN       36.6                26         36.612903  ...   
75756                51.0       36.6                27         36.935484  ...   

       days_below_30F  days_below_20F  days_below_10F  days_below_0F  \
0                   0               0               0              0   
1                   0               0               0              0   
2                   0               0               0              0   
3                   0               0               0              0   
4                   0               0               0              0   
...               ...             ...             ...            ...   
75752               6               0               0              0   
75753               6               0               0              0   
75754              26               2               0              0   
75755              26               2               0              0   
75756              17               1               0              0   

       days_above_80F  days_above_90F  days_above_100F  days_above_110F  \
0                  14               0                0                0   
1                  14               0                0                0   
2                  14               0                0                0   
3                  14               0                0                0   
4                  14               0                0                0   
...               ...             ...              ...              ...   
75752              25               3                0                0   
75753              25               3                0 

#### ++ Variables con varianza cercana a cero

Aqui deben encontrar las columnas numericas con desviacion cercana a cero, pero antes de ser normalizadas.

In [ ]:
dscr_train= train.select_dtypes( include= 'floating').describe()

In [ ]:
# Coeficiente de variacion
dscr_train.loc['CV', :]= dscr_train.apply( lambda x: x.loc['std']/x.loc['mean'] , axis=0)

In [ ]:
# Porcentaje NA
dscr_train.loc['Porc_NA', :]=train.select_dtypes( include= 'floating').isnull().mean()

In [ ]:
dscr_train

floor_area    year_built  energy_star_rating     ELEVATION  \
count    7.575700e+04  73914.000000        49048.000000  75757.000000   
mean     1.659839e+05   1952.465243           61.048605     39.506323   
std      2.468758e+05     32.613687           28.663683     60.656596   
min      9.430000e+02   1600.000000            0.000000     -6.400000   
25%      6.237900e+04   1927.000000           40.000000     11.900000   
50%      9.136700e+04   1951.000000           67.000000     25.000000   
75%      1.660000e+05   1977.000000           85.000000     42.700000   
max      6.385382e+06   2015.000000          100.000000   1924.500000   
CV       1.487348e+00      0.016704            0.469522      1.535364   
Porc_NA  0.000000e+00      0.024328            0.352561      0.000000   

         january_avg_temp  february_avg_temp  march_avg_temp  april_avg_temp  \
count        75757.000000       75757.000000    75757.000000    75757.000000   
mean            34.310468          35.526837       44.469292       53.784863   
std              6.996108           8.866697        6.657142        2.500473   
min             10.806452          13.250000       25.854839       40.280000   
25%             29.827586          31.625000       38.096774       52.333333   
50%             34.451613          34.107143       44.516129       53.333333   
75%             37.322581          40.879310       49.354839       54.766667   
max             64.758065          65.107143       69.758065       74.500000   
CV               0.203906           0.249577        0.149702        0.046490   
Porc_NA          0.000000           0.000000        0.000000        0.000000   

         may_avg_temp  june_avg_temp  ...  november_avg_temp  \
count    75757.000000   75757.000000  ...       75757.000000   
mean        63.742547      71.066384  ...          48.123133   
std          2.792919       2.861681  ...           4.094207   
min         46.950000      54.100000  ...          30.766667   
25%         62.790323      70.983333  ...          45.250000   
50%         62.887097      72.250000  ...          47.116667   
75%         65.096774      72.450000  ...          51.339286   
max         82.112903      89.550000  ...          67.416667   
CV           0.043816       0.040268  ...           0.085078   
Porc_NA      0.000000       0.000000  ...           0.000000   

         december_avg_temp  precipitation_inches  snowfall_inches  \
count         75757.000000          75757.000000     75757.000000   
mean             41.625569             42.430651        29.136379   
std               5.869386             10.592205        17.810894   
min              23.790323              0.000000         0.000000   
25%              38.532258             39.350000         9.600000   
50%              40.516129             42.170000        31.600000   
75%              43.435484             46.320000        49.100000   
max              61.790323            107.690000       127.300000   
CV                0.141004              0.249636         0.611294   
Porc_NA           0.000000              0.000000         0.000000   

             avg_temp  direction_max_wind_speed  direction_peak_wind_speed  \
count    75757.000000              34675.000000               33946.000000   
mean        56.176705                 66.552675                  62.779974   
std          2.241940                131.147834                 130.308106   
min         44.512329                  1.000000                   1.000000   
25%         55.147541                  1.000000                   1.000000   
50%         56.835616                  1.000000                   1.000000   
75%         57.215847                  1.000000                   1.000000   
max         77.245205                360.000000                 360.000000   
CV           0.039909                  1.970587                   2.075632   
Porc_NA      0.000000                  0.542287                   0.551909   

  

In [ ]:
# Aquí estoy usando El Coeficiente de Variacion: desviación dividida para la media, 
# pero uds pueden usar cualquier otro análisis para encontrar estas variables con poca variabilidad
umbral_CV= 0.05
dscr_train.loc[:, dscr_train.apply( lambda x: x.loc['CV']< umbral_CV , axis=0)].columns

Index(['year_built', 'april_avg_temp', 'may_avg_temp', 'june_avg_temp',
       'october_avg_temp', 'avg_temp'],
      dtype='object')

In [ ]:
# Ver las descriptivas de estas variables (podría servirnos para ver cambios en percentiles)
dscr_train.loc[:, dscr_train.apply( lambda x: x.loc['CV']< umbral_CV , axis=0)]

year_built  april_avg_temp  may_avg_temp  june_avg_temp  \
count    73914.000000    75757.000000  75757.000000   75757.000000   
mean      1952.465243       53.784863     63.742547      71.066384   
std         32.613687        2.500473      2.792919       2.861681   
min       1600.000000       40.280000     46.950000      54.100000   
25%       1927.000000       52.333333     62.790323      70.983333   
50%       1951.000000       53.333333     62.887097      72.250000   
75%       1977.000000       54.766667     65.096774      72.450000   
max       2015.000000       74.500000     82.112903      89.550000   
CV           0.016704        0.046490      0.043816       0.040268   
Porc_NA      0.024328        0.000000      0.000000       0.000000   

         october_avg_temp      avg_temp  
count        75757.000000  75757.000000  
mean            58.727038     56.176705  
std              2.641300      2.241940  
min             44.693548     44.512329  
25%             58.048387     55.147541  
50%             58.758065     56.835616  
75%             59.629032     57.215847  
max             80.741935     77.245205  
CV               0.044976      0.039909  
Porc_NA          0.000000      0.000000

In [ ]:
## Seleccionar columnas por nombre en el train (sólo para probar)
train.loc[ :, dscr_train.loc[:, dscr_train.apply( lambda x: x.loc['CV']< umbral_CV , axis=0)].columns ]

year_built  april_avg_temp  may_avg_temp  june_avg_temp  \
0          1942.0       55.500000     56.854839      60.500000   
1          1955.0       55.500000     56.854839      60.500000   
2          1951.0       55.500000     56.854839      60.500000   
3          1980.0       55.500000     56.854839      60.500000   
4          1985.0       55.500000     56.854839      60.500000   
...           ...             ...           ...            ...   
75752      1995.0       56.266667     59.193548      63.566667   
75753      1910.0       56.266667     59.193548      63.566667   
75754      1917.0       50.183333     52.145161      56.233333   
75755      2012.0       50.183333     52.145161      56.233333   
75756      1974.0       51.533333     53.887097      58.433333   

       october_avg_temp   avg_temp  
0             63.016129  56.972603  
1             63.016129  56.972603  
2             63.016129  56.972603  
3             63.016129  56.972603  
4             63.016129  56.972603  
...                 ...        ...  
75752         55.403226  55.147541  
75753         55.403226  55.147541  
75754         47.661290  47.911202  
75755         47.661290  47.911202  
75756         48.532258  49.127397  

[75757 rows x 6 columns]

**Debate**  
¿Está bien usar el coeficiente de variación? O prefieren usar rango intercuartil por ejemplo? O alguna otra medida que les diga cuán "poco variables" son los datos?... Y eliminarían todas las variables? Cuáles sí??

In [ ]:
#Grupo JAK2
#Si me parece bien el CV.

#### ++ Variables categoricas cuyo label mayor es cercano a 100%

Aquí deben desarrollar algo que permita encontrar cuál es el label con mayor frecuencia en cada variable categórica; luego ver el porcentaje de ese label. El resultado debería ser algo así:  

| Variable | Label | Porc |
| --- | --- | --- |
| Satisfacción | Bueno | 90 |
| País_Nacimi | Ecuad | 98 |

Lo que luego nos permitirá filtrar esas variable cuya primer categoría (en frecuencia) sea cercana a 100 (umbral por definir)


In [ ]:
#Grupo JAK2
train.head()

Year_Factor State_Factor building_class  \
0            1      State_1     Commercial   
1            1      State_1     Commercial   
2            1      State_1     Commercial   
3            1      State_1     Commercial   
4            1      State_1     Commercial   

                               facility_type  floor_area  year_built  \
0               Grocery_store_or_food_market     61242.0      1942.0   
1  Warehouse_Distribution_or_Shipping_center    274000.0      1955.0   
2                       Retail_Enclosed_mall    280025.0      1951.0   
3                  Education_Other_classroom     55325.0      1980.0   
4                  Warehouse_Nonrefrigerated     66000.0      1985.0   

   energy_star_rating  ELEVATION  january_min_temp  january_avg_temp  ...  \
0                11.0        2.4                36              50.5  ...   
1                45.0        1.8                36              50.5  ...   
2                97.0        1.8                36              50.5  ...   
3                46.0        1.8                36              50.5  ...   
4               100.0        2.4                36              50.5  ...   

   days_above_80F  days_above_90F  days_above_100F  days_above_110F  \
0              14               0                0                0   
1              14               0                0                0   
2              14               0                0                0   
3              14               0                0                0   
4              14               0                0                0   

   direction_max_wind_speed  direction_peak_wind_speed  max_wind_speed  \
0                       1.0                        1.0             1.0   
1                       1.0                        NaN             1.0   
2                       1.0                        NaN             1.0   
3                       1.0                        NaN             1.0   
4                       1.0                        1.0             1.0   

   days_with_fog    site_eui  id  
0            NaN  248.682615   0  
1           12.0   26.500150   1  
2           12.0   24.693619   2  
3           12.0   48.406926   3  
4            NaN    3.899395   4  

[5 rows x 64 columns]

In [ ]:
#Grupo JAK2
train["building_class"]

0         Commercial
1         Commercial
2         Commercial
3         Commercial
4         Commercial
            ...     
75752     Commercial
75753    Residential
75754     Commercial
75755     Commercial
75756    Residential
Name: building_class, Length: 75757, dtype: object

In [ ]:
#Grupo JAK2
agrup = train.groupby(['building_class'])['building_class'].count()
print(agrup)

building_class
Commercial     32199
Residential    43558
Name: building_class, dtype: int64


#### ++ Variables categoricas que deben unirse las categorias

En R esto lo hace automaticamente tidymodels (https://recipes.tidymodels.org/reference/step_other.html), en Python no es así y esto afecta mucho en términos de RAM al modelamiento (por la cantidad de variables dummy que se crean)

#### ++ Análisis simple de la variable Y

- Categórica? Cuántos niveles?
- Numérica? Cómo se distribuye? tiene Outliers?

#### Conclusiones

Aquí deben poner las conclusiones de todo el proceso de EDA:  
- ¿Qué variables se eliminan y porqué?
- ¿Qué variables se van a transformar, por ejemplo creando una nueva que sea binaria e indique que hubo missing_value
- ¿Existe alguna variables que amerite realizarle una transformación (cuadrática, logaritmica)?

--------------
--------------
--------------

### EDA Multivariado

Para encontrar 
- Relaciones entre variables como variables que pueden producir tener multicolinearidad o concurvidad, etc.
- Cómo se comportan las variables vs la variable Y 
- Existen valores aberrantes cuando se comparan variables (o pares de variables) vs nuestra Y

#### ++ Variables con alta correlación